src/string-view.h

```diff
@@ -0,0 +1,259 @@
+  string_view& operator=(const string_view&) noexcept = default;
+  string_view(const std::string& str) noexcept;
+  string_view(const char* str);
+  string_view(const char* str, size_type len);
+
+  // iterator support
+  const_iterator begin() const noexcept;
+  const_iterator end() const noexcept;
+  const_iterator cbegin() const noexcept;
+  const_iterator cend() const noexcept;
+  const_reverse_iterator rbegin() const noexcept;
+  const_reverse_iterator rend() const noexcept;
+  const_reverse_iterator crbegin() const noexcept;
+  const_reverse_iterator crend() const noexcept;
+
+  // capacity
+  size_type size() const noexcept;
+  size_type length() const noexcept;
+  size_type max_size() const noexcept;
+  bool empty() const noexcept;
+
+  // element access
+  const_reference operator[](size_type pos) const;
+  const_reference at(size_type pos) const;
+  const_reference front() const;
+  const_reference back() const;
+  const_pointer data() const noexcept;
+
+  // modifiers
+  void remove_prefix(size_type n);
+  void remove_suffix(size_type n);
+  void swap(string_view& s) noexcept;
+
+  // string operations
+  explicit operator std::string() const;
+  std::string to_string() const;
+  size_type copy(char* s, size_type n, size_type pos = 0) const;
+  string_view substr(size_type pos = 0, size_type n = npos) const;
+  int compare(string_view s) const noexcept;
+  int compare(size_type pos1, size_type n1, string_view s) const;
+  int compare(size_type pos1,
+              size_type n1,
+              string_view s,
+              size_type pos2,
+              size_type n2) const;
+  int compare(const char* s) const;
+  int compare(size_type pos1, size_type n1, const char* s) const;
+  int compare(size_type pos1, size_type n1, const char* s, size_type n2) const;
+  size_type find(string_view s, size_type pos = 0) const noexcept;
+  size_type find(char c, size_type pos = 0) const noexcept;
+  size_type find(const char* s, size_type pos, size_type n) const;
+  size_type find(const char* s, size_type pos = 0) const;
+  size_type rfind(string_view s, size_type pos = npos) const noexcept;
+  size_type rfind(char c, size_type pos = npos) const noexcept;
+  size_type rfind(const char* s, size_type pos, size_type n) const;
+  size_type rfind(const char* s, size_type pos = npos) const;
+  size_type find_first_of(string_view s, size_type pos = 0) const noexcept;
+  size_type find_first_of(char c, size_type pos = 0) const noexcept;
+  size_type find_first_of(const char* s, size_type pos, size_type n) const;
+  size_type find_first_of(const char* s, size_type pos = 0) const;
+  size_type find_last_of(string_view s, size_type pos = npos) const noexcept;
+  size_type find_last_of(char c, size_type pos = npos) const noexcept;
+  size_type find_last_of(const char* s, size_type pos, size_type n) const;
+  size_type find_last_of(const char* s, size_type pos = npos) const;
+
+  // TODO(binji): implement
+#if 0
```

Comment:
I figured it was useful to see what is not implemented from C++17 basic_string_view. I can remove it if you feel strongly